scipy/stats/tests/test_stats.py

```diff
@@ -537,7 +537,99 @@ def test_gh3014(self):
         # before, this would have risen a ValueError
         odds, pvalue = stats.fisher_exact([[1, 2], [9, 84419233]])
 
+class TestBarnardExact(object):
+    """Some tests to show that barnard_exact() works correctly.
+    """
+
+    def test_basic(self):
+        # results test against R's Barnard's Unconditional Test
+
+        barnard_exact = stats.barnard_exact
+        TX_obs, pvalue = barnard_exact([[7, 12], [8, 3]])
+        assert_almost_equal(
+            [TX_obs, pvalue],
+            [1.894338, 0.068153], decimal=5)
 
+        TX_obs, pvalue = barnard_exact(
+            [[43, 40], [10, 39]])
+        assert_almost_equal(
+            [TX_obs, pvalue],
+            [-3.55541, 0.000362832], decimal=5)
+
+        res = barnard_exact([[100, 2], [1000, 5]])[1]
+        assert_approx_equal(res, 0.13507, significant=2)
+        res = barnard_exact([[5, 1], [10, 10]])[1]
+        assert_approx_equal(res, 0.156278, significant=6)
+        res = barnard_exact([[5, 15], [20, 20]])[1]
+        assert_approx_equal(res, 0.0663635, significant=6)
+        res = barnard_exact([[5, 16], [20, 25]])[1]
+        assert_approx_equal(res, 0.116984, significant=6)
+        res = barnard_exact([[10, 5], [10, 1]])[1]
+        assert_approx_equal(res, 0.177537, significant=6)
+        res = barnard_exact([[5, 0], [1, 4]])[1]
+        assert_approx_equal(res, 0.0136719, significant=5)
+        res = barnard_exact([[0, 1], [3, 2]])[1]
+        assert_approx_equal(res, 0.509668, significant=5)
+        res = barnard_exact([[0, 2], [6, 4]])[1]
+        assert_approx_equal(res, 0.197019, significant=5)
+        res = barnard_exact([[2, 7], [8, 2]])
+        assert_approx_equal(res[1], 0.0192108, significant=5)
+        assert_approx_equal(res[0], 2.51847, significant=5)
```

Comment:
Why the `significant` changes?

Reply:
I understand why you do not retrieve same results than R. You do not use the same number of points than used in R function.

So we have two questions to answer:

1. Who's wrong?
   a. R is wrong. Then you can keep different significant values, but you have to explain why.
   b. Your code. Then you have to fix the same number of point.
2. How to handle maximization?
   a. As all found implementation found. By grid search as you do now.
   b. With a numeric scalar optimization (but the minimize function is not convex, therefore you should be very carefull).
   c. By a iterative grid search. (1). set a interval [0,1], (2). You do a grid search in the intervall with a small number of point, e.g. 100, (3) you find the max, (4) set the interval as [point before max, point after max], (5) go to (2). With a very small number of iteration, e.g. 3, you achieve a very good precision, and you can provide the iteration number in args.

My opinion is 1->a, and 2->c.